compiler/rustc_macros/src/diagnostics/diagnostic.rs

```diff
@@ -152,6 +156,41 @@ impl<'a> LintDiagnosticDerive<'a> {
             }
         });
 
+        let span = Self::KIND.each_variant(&mut structure, |_, variant| {
```

Comment:
Initially I tried to implement this in a DRY way reusing existing functions from `diagnostic_builder` but in the end I opted for just doing it the straightforward way because it wasn't entirely obvious to me how I should refactor `diagnostic_builder` to suit my needs.